project-api/project-api-test/src/test/java/io/ballerina/projects/test/PreReleaseVersionResolutionTests.java

```diff
@@ -0,0 +1,121 @@
+import org.testng.annotations.Test;
+
+import java.io.PrintStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collection;
+
+/**
+ * Contains cases to test pre-release version resolution logic.
+ *
+ * @since 2.0.0
+ */
+public class PreReleaseVersionResolutionTests {
+    private static final Path RESOURCE_DIRECTORY = Paths.get(
+            "src/test/resources/projects_for_resolution_tests").toAbsolutePath();
+    static final PrintStream OUT = System.out;
+
+    @BeforeSuite
+    public void init() {
+        BCompileUtil.compileAndCacheBala("projects_for_resolution_tests/package_unstable_k_alpha");
+        BCompileUtil.compileAndCacheBala("projects_for_resolution_tests/package_unstable_k_beta");
+        BCompileUtil.compileAndCacheBala("projects_for_resolution_tests/package_unstable_k_GA");
+        BCompileUtil.compileAndCacheBala("projects_for_resolution_tests/package_l_with_unstable_dep");
+    }
+
+    @Test(description = "tests projects with pre-release versions")
+    public void testProjectWithAPreReleaseVersion() {
+        // package_unstable_k_alpha has no dependencies and its version is 1.1.0-alpha
+        Path projectDirPath = RESOURCE_DIRECTORY.resolve("package_unstable_k_alpha");
+        BuildProject buildProject = BuildProject.load(projectDirPath);
+        PackageCompilation compilation = buildProject.currentPackage().getCompilation();
+
+        // Check whether there are any diagnostics
+        DiagnosticResult diagnosticResult = compilation.diagnosticResult();
+        diagnosticResult.errors().forEach(OUT::println);
+        Assert.assertEquals(diagnosticResult.diagnosticCount(), 0, "Unexpected compilation diagnostics");
+
+        // Check direct package dependencies
+        Assert.assertEquals(buildProject.currentPackage().packageDependencies().size(), 0,
+                "Unexpected number of dependencies");
+    }
+
+    @Test(description = "Two pre-release versions (specified by the user) in the dependency graph",
+            expectedExceptions = ProjectException.class,
+            expectedExceptionsMessageRegExp = "Two incompatible versions exist in the dependency graph: " +
+                    "samjs/package_k versions: 1.1.0-alpha, 1.1.0-beta")
+    public void testTwoIncompatiblePreReleaseVersionsInDependencyGraph() {
+        // package_m_with_unstable_dep --> package_l (2.1.0) --> package_k (1.1.0-beta)
+        // package_l (2.1.0) --> package_k (1.1.0-alpha)
+        // This test case should fail. There are two incompatible versions of package-k in the dependency graph of
+        //  package_m_with_unstable_dep.
+        Path projectDirPath = RESOURCE_DIRECTORY.resolve("package_m_with_unstable_dep");
+        BuildProject buildProject = BuildProject.load(projectDirPath);
+        buildProject.currentPackage().getResolution();
+    }
+
+    @Test(description = "Two pre-release versions in the dependency graph, one is specified by the user, " +
```

Comment:
Do we have a case for two major versions in the dependency graph one is specified by user other has come as the latest ?

Reply:
There are no test cases like that I will add that as a separate PR. I will that as a separate PR.